gnucash/gnome/assistant-stock-transaction.cpp

```diff
@@ -0,0 +1,1148 @@
+    auto gae = gnc_amount_edit_new ();
+    gnc_amount_edit_set_evaluate_on_enter (GNC_AMOUNT_EDIT (gae), TRUE);
+    gnc_amount_edit_set_print_info (GNC_AMOUNT_EDIT (gae), info);
+    gtk_grid_attach (GTK_GRID(table), gae, 1, row, 1, 1);
+    gtk_widget_show (gae);
+    gnc_amount_edit_make_mnemonic_target (GNC_AMOUNT_EDIT (gae), label);
+    return gae;
+}
+
+static GtkWidget * create_date (GtkBuilder *builder, guint row,
+                                const gchar *date_label, const gchar *table_label)
+{
+    auto date = gnc_date_edit_new (gnc_time (NULL), FALSE, FALSE);
+    auto label = get_widget (builder, date_label);
+    gtk_grid_attach (GTK_GRID(get_widget (builder, table_label)), date, 1, row, 1, 1);
+    gtk_widget_show (date);
+    gnc_date_make_mnemonic_target (GNC_DATE_EDIT(date), label);
+    return date;
+}
+
+static GtkWidget * get_treeview (GtkBuilder *builder, const gchar *treeview_label)
+{
+    auto view = GTK_TREE_VIEW (get_widget (builder, "transaction_view"));
+    gtk_tree_view_set_grid_lines (GTK_TREE_VIEW(view), gnc_tree_view_get_grid_lines_pref ());
+
+    auto store = gtk_list_store_new (NUM_SPLIT_COLS, G_TYPE_STRING, G_TYPE_STRING,
+                                     G_TYPE_STRING, G_TYPE_STRING, G_TYPE_STRING);
+    gtk_tree_view_set_model(view, GTK_TREE_MODEL(store));
+    g_object_unref(store);
+
+    auto renderer = gtk_cell_renderer_text_new();
+    auto column = gtk_tree_view_column_new_with_attributes
+        (_("Account"), renderer, "text", SPLIT_COL_ACCOUNT, nullptr);
+    gtk_tree_view_append_column(view, column);
+
+    renderer = gtk_cell_renderer_text_new();
+    column = gtk_tree_view_column_new_with_attributes
+        (_("Memo"), renderer, "text", SPLIT_COL_MEMO, nullptr);
+    gtk_tree_view_append_column(view, column);
+
+    renderer = gtk_cell_renderer_text_new();
+    column = gtk_tree_view_column_new_with_attributes
+        (_("Debit"), renderer, "text", SPLIT_COL_DEBIT, nullptr);
+    gtk_tree_view_append_column(view, column);
+
+    renderer = gtk_cell_renderer_text_new();
+    column = gtk_tree_view_column_new_with_attributes
+        (_("Credit"), renderer, "text", SPLIT_COL_CREDIT, nullptr);
+    gtk_tree_view_append_column(view, column);
+
+    return GTK_WIDGET (view);
+}
+
+static GtkWidget *
+stock_assistant_create (StockTransactionInfo *info)
+{
+    auto builder = gtk_builder_new();
+    gnc_builder_add_from_file  (builder , "assistant-stock-transaction.glade", "stock_transaction_assistant");
+    info->window = get_widget (builder, "stock_transaction_assistant");
+
+    // Set the name for this assistant so it can be easily manipulated with css
+    gtk_widget_set_name (GTK_WIDGET(info->window), "gnc-id-assistant-stock-transaction");
+
+    auto balance = xaccAccountGetBalance (info->acct);
+    info->txn_types = gnc_numeric_zero_p (balance) ? starting_types
+        : gnc_numeric_positive_p (balance) ? long_types
+        : short_types;
```

Comment:
Well, if you don't want a copy, don't tell it to copy.

Making `txn_types` a `TxnTypeInfoVec&` should prevent the copying and making the `std::string`->`const char*` and `std::vector`->`constexpr std::array` changes I suggested up top should fix a lot of the allocating. You can safely use `std::string_view`s if you need to use `std::string` functionality on those `const char*`s, since they'll be part of the `constexpr std::array<TxnTypeInfo>`s you don't have to worry about them going anywhere.